### PR TITLE
payment: Fix deserialization for refund+item adj. See #1453

### PR DESCRIPTION
**Pull request by sbrossie**

### CODE REVIEW

File: payment/src/main/java/org/killbill/billing/payment/invoice/InvoicePaymentControlPluginApi.java [hunk 1]
```diff
@@ -459,7 +460,37 @@ private PriorPaymentControlResult getPluginCreditResult(final PaymentControlCont
...
+            }
+
+            final Object v = m.get(k);
+            BigDecimal val;
+            if (v instanceof BigDecimal) {
+                val = (BigDecimal) v;
+            } else if (v instanceof String) {
+                val = new BigDecimal((String) v);
+            } else if (v instanceof Integer) {
+                val = BigDecimal.valueOf((Integer) v);
```
Comment (pierre) [inline payment/src/main/java/org/killbill/billing/payment/invoice/InvoicePaymentControlPluginApi.java:484]: Not sure this does what we want. Would it be easier to do something like?
```
try {
  val = new BigDecimal(v.toString());
} catch () { // throw }
```

File: payment/src/main/java/org/killbill/billing/payment/invoice/InvoicePaymentControlPluginApi.java [hunk 2]
```diff
@@ -459,7 +460,37 @@ private PriorPaymentControlResult getPluginCreditResult(final PaymentControlCont
...
         }
-        return (Map<UUID, BigDecimal>) prop.getValue();
+        // The deserialization may not recreate the map we expect, i.e Map<UUID, BigDecimal>, so we convert each key/value by hand
+        // See https://github.com/killbill/killbill/issues/1453
+        final Map<UUID, BigDecimal> res = new HashMap<>();
+        final Map m = (Map) prop.getValue();
+        for (final Object k : m.keySet()) {
+            UUID uuid;
+            if (k instanceof String) {
+                uuid = UUID.fromString((String) k);
```
Comment (pierre) [inline payment/src/main/java/org/killbill/billing/payment/invoice/InvoicePaymentControlPluginApi.java:470]: You could maybe be a bit more generic:

```
try {
  uuid = UUID.fromString(v.toString());
} catch () { // throw }
```

That way, we'd handle other types as well (e.g. `CharSequence`).

File: payment/src/main/java/org/killbill/billing/payment/invoice/InvoicePaymentControlPluginApi.java (hunk 1 above)
Comment (sbrossie) [reply payment/src/main/java/org/killbill/billing/payment/invoice/InvoicePaymentControlPluginApi.java:484]: Actually, if the amount is an integer, it gets deserialized as an `Integer` and so I need this `else if` for this.

Comment (sbrossie) [reply payment/src/main/java/org/killbill/billing/payment/invoice/InvoicePaymentControlPluginApi.java]: Fixed in https://github.com/killbill/killbill/pull/1456/commits/a915db3c09ad7c6c384f7c3fe8e0848f36eda12b

File: payment/src/main/java/org/killbill/billing/payment/invoice/InvoicePaymentControlPluginApi.java (hunk 2 above)
Comment (sbrossie) [reply payment/src/main/java/org/killbill/billing/payment/invoice/InvoicePaymentControlPluginApi.java:470]: The only use case we really have is one when we ned to convert String -> UUID - i'd rather keep the code as it is.

**Pull request merged**